synapse/rest/client/v1/room.py

```diff
@@ -55,17 +55,11 @@ def register(self, http_server):
 
     @defer.inlineCallbacks
     def on_PUT(self, request, txn_id):
-        try:
-            defer.returnValue(
-                self.txns.get_client_transaction(request, txn_id)
-            )
-        except KeyError:
-            pass
-
-        response = yield self.on_POST(request)
-
-        self.txns.store_client_transaction(request, txn_id, response)
-        defer.returnValue(response)
+        observable = self.txns.fetch_or_execute_request(
+            request, self.on_POST, request
+        )
+        res = yield observable.observe()
```

Comment:
Ah, I'd move this `.observe()` up into the actual cache to make things neater:

```python
def on_PUT(self, request, txn_id):
    return self.txns.fetch_or_execute_request(
        request, self.on_POST, request
    )
````